proposals/lip-0052.md

```diff
@@ -362,712 +379,1832 @@ The module provides the following commands to modify the NFT store.
-    or length(senderAddress) != ADDRESS_LENGTH
-    or length(recipientAddress) !=  ADDRESS_LENGTH
-    or length(receivedAttributes) > MAX_BYTE_LENGTH_ATTRIBUTES
-    or (chainID == ownChainID
-        and NFTStore(nftID).owner != uint32(sendingChainID))
-    or (chainID == sendingChainID
-        and NFTStore(nftID) exists)):  
-    if  (ccm.fee >= MIN_RETURN_FEE * length(ccm)
-        and ccm.status == CCM_STATUS_OK):
-        interoperability.error(ccm, CCM_STATUS_NFT_PROTOCOL_VIOLATION)
-    interoperability.terminateChain(sendingChainID)
-    stop ccm execution
-
-if chainID == ownChainID:
-    oldAttributes = NFTStore(nftID).attributes
-    if ccm.status == CCM_STATUS_OK:
-        newAttributes = getNewAttributes(nftID.collection,
-                                         oldAttributes,
-                                         receivedAttributes)
-        newRecipientAddress = recipientAddress
-    else:
-        newAttributes = oldAttributes
-        newRecipientAddress = senderAddress
-
-    nftID = canonicalNFTID(nftID)
-    NFTStore(nftID).owner = newRecipienAddress
-    NFTStore(nftID).attributes = newAttributes
-    createUserEntry(newRecipientAddress, nftID)
-
-else: # chainID == sendingChainID
-    if NFTSupported(nftID) == FALSE:
-        if (ccm.fee >= MIN_RETURN_FEE * length(ccm)
-            and ccm.status == CCM_STATUS_OK):
-            interoperability.error(ccm, CCM_STATUS_NFT_NOT_SUPPORTED)
-        stop ccm execution
-    if ccm.status == CCM_STATUS_OK:
-        createNFTEntry(nftID,
-                       recipientAddress,
-                       receivedAttributes)
+def execute(
+    ccu: Transaction,
+    ccm: CCM
+) -> None:
+
+    nftID = ccm.params.nftID
+    senderAddress = ccm.params.senderAddress
+    recipientAddress = ccm.params.recipientAddress
+    data = ccm.params.data
+    sendingChainID = ccm.sendingChainID
+
+    if isNFTSupported(nftID) == False:
+        emitPersistentEvent(
+            module = MODULE_NAME_NFT,
+            name = EVENT_NAME_CCM_TRANSFER,
+            data = {
+                "senderAddress": senderAddress,
+                "recipientAddress": recipientAddress,
+                "nftID": nftID,
+                "receivingChainID": OWN_CHAIN_ID,
+                "result": RESULT_NFT_NOT_SUPPORTED
+            },
+            topics = [senderAddress, recipientAddress]
+        )
+        raise Exception("Non-supported NFT.")
+
+    if ccm.status != CCM_STATUS_OK:
+        recipientAddress = senderAddress
+
+    if NFTStore(nftID) does not exist:
+        createNFTEntry(recipientAddress, nftID)
         createUserEntry(recipientAddress, nftID)
-    else:
-        createNFTEntry(nftID,
-                       senderAddress,
-                       receivedAttributes)
-        createUserEntry(senderAddress, nftID)
+
+    if getChainID(nftID) == OWN_CHAIN_ID:
+        NFTStore(nftID).owner = recipientAddress
+        escrowEntry(sendingChainID, nftID).escrow = False
```

Comment:
I think we should rather delete the entry for consistency. If the entry does not exist, it is not escrowed, otherwise we also have to check the value.